tests/unit/test_jobs.py

```diff
@@ -0,0 +1,33 @@
+from catkin_tools.jobs.utils import merge_envs
+
+
+def test_merge_envs_basic():
+    job_env = { 'PATH': '/usr/local/bin:/usr/bin', 'FOO': 'foo' }
+
+    merge_envs(job_env, [
+        { 'PATH': '/usr/local/bin:/bar/baz/bin' },
+        { 'BAR': 'bar' } ])
+
+    # Validate that the known path was not moved from the existing order, and the unfamiliar
+    # path was correctly prepended.
+    assert job_env['PATH'] == '/bar/baz/bin:/usr/local/bin:/usr/bin'
```

Comment:
I see what you mean now. That's unresolvable I think as-is because each package brings a set of underlays which cannot be deduplicated at the "underlay" level. This might be something that can be addressed by a feature in ament (hopefully a back portable thing). Each package has two setup files, `setup.*sh` and `localsetup.*sh`. The latter only modifies the environment for that package. That means you can start with an underlay and know what each package brings to the table separately from their dependencies.

Anyways if you can remove this code path from the non-isolated case then this will have to be good enough I think.